# Lab 4 — Reading and Writing Data Files

**Course:** Python for AI & Data  
**Analyst:** *Natalia Amaya*  
**Date:** *4/20/2026*

---

## The BeanScene Scenario

Sarah has exported BeanScene's menu data to a CSV file and stored her performance thresholds in a JSON config file. Your job is to load both files, run the menu analysis, and save the results back to disk.

Work from top to bottom. Run every cell as you go.


---
## Task 1 — Confirm Your Environment

Before opening any files, confirm:
1. Your working directory is the repo root (not inside `notebooks/`)
2. The data files exist at the expected paths

If the paths show `False`, close JupyterLab, `cd` to the repo root, and relaunch.


In [7]:
from pathlib import Path

print("Working directory:", Path.cwd())

# Added this to go up one level from notebooks to project root
PROJECT_ROOT = Path.cwd().parent

RAW_DIR = PROJECT_ROOT / "data" / "raw"

menu_path = RAW_DIR / "beanscene_menu.csv"
config_path = RAW_DIR / "beanscene_config.json"

print(menu_path.exists())
print(config_path.exists())

Working directory: C:\Users\Natalia Briscoe\desktop\Lab 4\lab-04-reading-and-writing-data-files\notebooks
True
True


---
## Task 2 — Load the Menu from CSV

Read `beanscene_menu.csv` into a list of dictionaries.

After loading:
- Convert `price` to `float` and `units_sold` to `int` for each row
- `name` and `category` should remain strings

Print the first item and check the types to confirm everything loaded correctly.

**Hint:** Use `csv.DictReader` — it reads each row as a dictionary using the header row as keys.


In [ ]:
import csv

menu = []

# Open the CSV file and read into menu (list of dicts with correct types)


# Print the first item and verify types
print(menu[0])
print("price type:", type(menu[0]["price"]))
print("units_sold type:", type(menu[0]["units_sold"]))
print(f"Total items loaded: {len(menu)}")


In [12]:
import csv
from pathlib import Path

menu_path = Path.cwd().parent / "data" / "raw" / "beanscene_menu.csv"

menu = []

with open(menu_path, newline="", encoding="utf-8") as file:
    reader = csv.DictReader(file)
    for row in reader:
        menu.append(row)

In [14]:
print(menu[0])
print("price type:", type(menu[0]["price"]))
print("units_sold type:", type(menu[0]["units_sold"]))
print(f"Total items loaded: {len(menu)}")


{'name': 'Espresso', 'price': '3.50', 'units_sold': '42', 'category': 'Hot Drinks'}
price type: <class 'str'>
units_sold type: <class 'str'>
Total items loaded: 8


In [22]:
menu = []

In [23]:
with open(menu_path, newline="", encoding="utf-8") as file:
    reader = csv.DictReader(file)
    for row in reader:
        # Convert types
        row["price"] = float(row["price"])
        row["units_sold"] = int(row["units_sold"])

        # Keep name and category as strings (already are, but explicit is fine)
        row["name"] = str(row["name"])
        row["category"] = str(row["category"])

        menu.append(row)

# Print first item + type checks
print(menu[0])
print("price type:", type(menu[0]["price"]))
print("units_sold type:", type(menu[0]["units_sold"]))
print(f"Total items loaded: {len(menu)}")

{'name': 'Espresso', 'price': 3.5, 'units_sold': 42, 'category': 'Hot Drinks'}
price type: <class 'float'>
units_sold type: <class 'int'>
Total items loaded: 8


---
## Task 3 — Load the Config from JSON

Read `beanscene_config.json` into a Python dictionary.

After loading, print:
- The café name
- The `high` performance threshold
- The `low_performer_threshold`

**Note:** Unlike CSV, JSON numbers load as actual Python numbers — no conversion needed.


In [ ]:
import json

config = {}

# Open the JSON file and load into config


# Print key values from the config
print("Café name:", config["cafe_name"])
print("High threshold:", config["performance_thresholds"]["high"])
print("Low threshold:", config["low_performer_threshold"])


In [27]:
import json

with open(config_path) as f:
    config = json.load(f)

In [29]:
print("Café name:", config["cafe_name"])
print("High threshold:", config["performance_thresholds"]["high"])
print("Low threshold:", config["low_performer_threshold"])

Café name: BeanScene
High threshold: 250.0
Low threshold: 150.0


---
## Task 4 — Analyse the Menu

Using the loaded `menu` list and `config` dictionary:

1. Calculate revenue for each item (`price * units_sold`)
2. Classify each item as `"High"`, `"Medium"`, or `"Low"` using the thresholds from `config` — **not hardcoded numbers**
3. Store the results in a new list called `results` — each item should be a dict with keys: `name`, `revenue`, `tier`
4. Print the results in a readable format

**Tip:** You can reuse the `classify_item` logic from Lab 3 — either copy the function or rewrite it here.


In [ ]:
# Extract thresholds from config
high_threshold = config["performance_thresholds"]["high"]
low_threshold = config["low_performer_threshold"]

results = []

# Loop over menu, calculate revenue, classify each item, append to results


# Print results


In [36]:
high_threshold = config["performance_thresholds"]["high"]
low_threshold = config["low_performer_threshold"]

results = []

for item in menu:
    revenue = item["price"] * item["units_sold"]

    if revenue >= high_threshold:
        tier = "High"
    elif revenue >= low_threshold:
        tier = "Medium"
    else:
        tier = "Low"

    results.append({
        "name": item["name"],
        "revenue": revenue,
        "tier": tier
    })

In [39]:
for item in results:
    print(f"{item['name']} | ${item['revenue']:.2f} | {item['tier']}")

Espresso | $147.00 | Low
Latte | $289.75 | High
Cappuccino | $171.00 | Medium
Cold Brew | $145.00 | Low
Flat White | $199.75 | Medium
Matcha Latte | $181.50 | Medium
Iced Americano | $220.00 | Medium
Chai Latte | $194.75 | Medium


---
## Task 5 — Save Results to CSV

Write your `results` list to `data/processed/menu_results.csv`.

The output file should have three columns: `name`, `revenue`, `tier`.

Make sure `data/processed/` exists before writing (create it if needed).


In [ ]:
# Create the processed directory if it doesn't exist
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

output_csv = PROCESSED_DIR / "menu_results.csv"

# Write results to CSV


print("Saved:", output_csv)


In [44]:
import csv
from pathlib import Path

# Resetting the project root again
PROJECT_ROOT = Path.cwd().parent

output_csv = PROJECT_ROOT / "data" / "processed" / "menu_results.csv"

with open(output_csv, "w", newline="", encoding="utf-8") as file:
    writer = csv.DictWriter(file, fieldnames=["name", "revenue", "tier"])
    writer.writeheader()
    writer.writerows(results)

print("Saved:", output_csv)

Saved: C:\Users\Natalia Briscoe\desktop\Lab 4\lab-04-reading-and-writing-data-files\data\processed\menu_results.csv


---
## Task 6 — Save a Summary to JSON

Build a summary dictionary and write it to `data/processed/weekly_summary.json`.

Include at minimum:
- `cafe_name` (from config)
- `week` (from config: `analysis_week`)
- `total_revenue` (sum of all revenues)
- `items_analysed` (number of items)
- `best_seller` (name of the item with highest revenue)

Use `indent=2` when writing so the file is human-readable.


In [ ]:
output_json = PROCESSED_DIR / "weekly_summary.json"

# Build the summary dictionary
summary = {
    # fill in the keys and values here
}

# Write to JSON


print("Saved:", output_json)


In [50]:
import json
from pathlib import Path


total_revenue = sum(item["revenue"] for item in results)
items_analysed = len(results)

best_seller_item = max(results, key=lambda x: x["revenue"])
best_seller = best_seller_item["name"]


summary = {
    "cafe_name": config["cafe_name"],
    "week": config["analysis_week"],
    "total_revenue": total_revenue,
    "items_analysed": items_analysed,
    "best_seller": best_seller
}


output_json = Path.cwd().parent / "data" / "processed" / "weekly_summary.json"


with open(output_json, "w", encoding="utf-8") as file:
    json.dump(summary, file, indent=2)

print("Saved", output_json)

Saved C:\Users\Natalia Briscoe\desktop\Lab 4\lab-04-reading-and-writing-data-files\data\processed\weekly_summary.json


---
## Task 7 — Verify Your Outputs

Re-load both files you just wrote and print their contents to confirm they were saved correctly.

This is a professional habit: always verify your outputs — don't assume they're correct.


In [46]:
# Re-load menu_results.csv and print all rows
print("=== menu_results.csv ===")


# Re-load weekly_summary.json and print
print("\n=== weekly_summary.json ===")


=== menu_results.csv ===

=== weekly_summary.json ===


In [54]:
import json
from pathlib import Path

json_file = Path.cwd().parent / "data" / "processed" / "weekly_summary.json"

with open(json_file, "r", encoding="utf-8") as file:
    summary = json.load(file)

print("\n=== weekly_summary.json ===")
print(summary)


=== weekly_summary.json ===
{'cafe_name': 'BeanScene', 'week': '2024-W04', 'total_revenue': 1548.75, 'items_analysed': 8, 'best_seller': 'Latte'}


In [53]:
import csv
from pathlib import Path

csv_file = Path.cwd().parent / "data" / "processed" / "menu_results.csv"

with open(csv_file, newline="", encoding="utf-8") as file:
    reader = csv.DictReader(file)
    menu_results_loaded = list(reader)

print("=== menu_results.csv ===")
print(menu_results_loaded[:3])

=== menu_results.csv ===
[{'{': '  "cafe_name": "BeanScene"', None: ['']}, {'{': '  "week": "2024-W04"', None: ['']}, {'{': '  "total_revenue": 1548.75', None: ['']}]


---
## ⭐ Optional Advanced Tasks

### Optional 1 — Low Performers File
Write a separate CSV to `data/processed/low_performers.csv` containing only the items classified as `"Low"`. Then look up append mode (`"a"`) and add a final row with the count of low performers.

### Optional 2 — Config-Driven Thresholds
Confirm your analysis is fully config-driven: open `data/raw/beanscene_config.json`, change the `high` threshold from `250` to `200`, save the file, and re-run Task 4. Verify the tiers change without editing any Python code. Then change it back.

### Optional 3 — Category Summary
Group the menu items by `category` and write a file `data/processed/category_summary.json` showing total revenue per category. (Hint: use a dictionary to accumulate revenue per category as you loop.)


In [ ]:
# Optional tasks — work here
